src/fhirpath.js

```diff
@@ -184,9 +185,24 @@ engine.invocationTable = {
 };
 
 engine.InvocationExpression = function(ctx, parentData, node) {
-  return node.children.reduce(function(acc, ch) {
+  let oldVars = undefined;
+  if (node.children.length > 0) {
+    // if the first child is a defineVariable function, stash the vars object
+    // so that it can be restored after the invocation is complete
+    if (node.children[0].type === 'TermExpression' && node.children[0].text.startsWith('defineVariable')) {
+      oldVars = ctx.vars;
+      // Clone a new vars object with all the current values in it
+      ctx.vars = Object.assign({}, ctx.vars);
```

Comment:
Do you mean like this?
``` javascript
engine.InvocationExpression = function(ctx, parentData, node) {
  let ctxForChildren = ctx;
  if (node.children.length > 0) {
    // if the first child is a defineVariable function, stash the vars object
    // so that it can be restored after the invocation is complete
    if (node.children[0].type === 'TermExpression' && node.children[0].text.startsWith('defineVariable')) {
      // shallow copy the context (and also shallow copy the vars)
      ctxForChildren = {...ctx, vars: {...ctx.vars}};
    }
  }
  return node.children.reduce(function(acc, ch) {
    return engine.doEval(ctxForChildren, acc, ch);
  }, parentData);
};
```

Reply:
Yes, I meant something like this.
But now I noticed some errors. I believe "defineVariable" could not be the first child. Also, we might have several  "defineVariable". Here are examples of tests that should pass, but do not pass:

```js
  it("defineVariable() could not be the first child", () => {
    // test with a variable that is not in the context that should throw an error
    let expr = `Patient.name.defineVariable('fam', first()).active | Patient.name.defineVariable('fam2', skip(1).first()).select(%fam.given)`;
    expect(() => {
      fhirpath.evaluate(input.patientExample, expr, r4_model); 
    }).toThrowError("Attempting to access an undefined environment variable: fam");
  });

  it("sequence of variable definitions", () => {
    let expr = `Patient.name.defineVariable('fam', first()).active | Patient.name.defineVariable('fam2', skip(1).first()).defineVariable('fam3', %fam2+%fam2).select(%fam3.given)`;
    expect(fhirpath.evaluate(input.patientExample, expr, r4_model))
      .toStrictEqual([true, "JimJim"]);
  });
```